cmd/kube-proxy/app/server.go

```diff
@@ -1086,22 +1098,21 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
 
 // getNodeIP returns IPs for the node with the provided name.  If
 // required, it will wait for the node to be created.
-func getNodeIPs(ctx context.Context, client clientset.Interface, name string) []net.IP {
+func getNodeIPs(ctx context.Context, nodeLister corelisters.NodeLister, name string) []net.IP {
 	logger := klog.FromContext(ctx)
 	var nodeIPs []net.IP
-	backoff := wait.Backoff{
-		Steps:    6,
-		Duration: 1 * time.Second,
-		Factor:   2.0,
-		Jitter:   0.2,
-	}
 
-	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
-		node, err := client.CoreV1().Nodes().Get(ctx, name, metav1.GetOptions{})
+	err := wait.PollUntilContextCancel(ctx, 1*time.Second, true, func(context.Context) (bool, error) {
+		node, err := nodeLister.Get(name)
 		if err != nil {
 			logger.Error(err, "Failed to retrieve node info")
 			return false, nil
 		}
+		// don't consider the node if is going to be deleted and keep waiting
+		if !node.DeletionTimestamp.IsZero() {
+			logger.Info("Node is being deleted", "node", node.Name)
+			return false, nil
+		}
```

Comment:
Thinking about the big picture of kube-proxy Node handling, we don't really deal well with the case where the Node is deleted and then not created (but kube-proxy is still running); we will crashloop in that case. Perhaps it would be better for kube-proxy to ignore Node deletion and _only_ restart on Node recreation? (In which case, we can just ignore DeletionTimestamp)

Reply:
we have very strong requirements for node startup and readiness, if we wait for crashing the delay will be greater than waiting here ..

Reply:
OK, but do we have strong requirements for "the Node object is deleted and then immediately recreated"?

It seems weird to have kube-proxy have to specifically distinguish the case of "the Node is in the process of being deleted but hasn't quite been deleted yet". Someone's abstraction is leaking all over the place.

Reply:
It's an edge case that happens because of two factor:
1. The VM is quickly terminated and recreated O(seconds):  Spot VM, preempted VMs or kind of VM migration situation (not live of course as the system is restarted). Common in cloud providers, at least common in GCE but afaik also AWS does something similar 
2. Run kube-proxy as a static pod, instead of a daemon set, so you kproxy on the node is tied to the node version 

I don't know now from the top of my mind of this deletion timestamp was me being paranoid or part of the bug, but with this setup you'll face the race eventually and your kproxy will get values from the old node object

Reply:
> 1. The VM is quickly terminated and recreated

With the Node object being deleted or not?

- **Node is deleted, then VM is restarted**: This is no different from the initial Node creation case; the VM will come up, kubelet and kube-proxy will be started, and kube-proxy has to wait for kubelet to create a new Node object. (So this already works fine.)
- **Node not deleted, Node not updated after VM restart**: Again, this already works fine; the correct Node will already exist when kube-proxy restarts.
- **Node not deleted, Kubelet updates it after VM restart**: (eg, because the VM gets a different IP after rebooting). In this case, it's possible that kube-proxy will start up and read the "old" Node, and then eventually kubelet will update the Node object and kube-proxy will have to restart (at which point it basically becomes the previous case: when kube-proxy restarts, the correct Node object will already exist).
- **VM restarted, then Node deleted**: In this case, kube-proxy will (potentially) start up with the "old" Node, as above, until it notices the Node deletion and restarts. If we make it so that (a) kube-proxy exits as soon as it sees a DeletionTimestamp on the Node rather than waiting for the deletion event, and (b) kube-proxy ignores a Node with a DeletionTimestamp at startup time, then we potentially get kube-proxy back up and running with correct configuration faster than we would have if we didn't do both (a) and (b).

So, yes, handling DeletionTimestamp can help in one case, but what would help even more would be if the infra made sure that if there's any chance the Node will need to be deleted or updated after the VM restart, then it should delete the Node before restarting...

> Run kube-proxy as a static pod, instead of a daemon set, so you kproxy on the node is tied to the node version

But what are the scenarios with that that would require the DeletionTimestamp check?